Blockzilla/PhotonActionSheet.swift

```diff
@@ -0,0 +1,596 @@
+            closeButton.snp.makeConstraints { make in
+                make.centerX.equalTo(self.view.snp.centerX)
+                make.width.equalTo(width)
+                make.height.equalTo(PhotonActionSheetUX.CloseButtonHeight)
+                make.bottom.equalTo(self.view.safeAreaLayoutGuide.snp.bottom).inset(PhotonActionSheetUX.Padding)
+            }
+        }
+        
+        if style == .popover {
+            self.actions = actions.map({ $0.reversed() }).reversed()
+            tableView.snp.makeConstraints { make in
+                make.edges.equalTo(self.view)
+            }
+        } else {
+            tableView.snp.makeConstraints { make in
+                make.centerX.equalTo(self.view.snp.centerX)
+                switch style {
+                case .bottom, .popover:
+                    make.bottom.equalTo(closeButton.snp.top).offset(-PhotonActionSheetUX.Padding)
+                case .centered:
+                    make.centerY.equalTo(self.view.snp.centerY)
+                }
+                make.width.equalTo(width)
+            }
+        }
+        tableView.alpha = 0.9
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        
+        tableView.bounces = false
+        tableView.delegate = self
+        tableView.dataSource = self
+        tableView.keyboardDismissMode = .onDrag
+        tableView.register(PhotonActionSheetCell.self, forCellReuseIdentifier: PhotonActionSheetUX.CellName)
+        tableView.register(PhotonActionSheetTitleHeaderView.self, forHeaderFooterViewReuseIdentifier: PhotonActionSheetUX.TitleHeaderName)
+        tableView.register(PhotonActionSheetSeparator.self, forHeaderFooterViewReuseIdentifier: "SeparatorSectionHeader")
+        tableView.register(UITableViewHeaderFooterView.self, forHeaderFooterViewReuseIdentifier: "EmptyHeader")
+        tableView.estimatedRowHeight = PhotonActionSheetUX.RowHeight
+        tableView.estimatedSectionFooterHeight = PhotonActionSheetUX.HeaderFooterHeight
+        // When the menu style is centered the header is much bigger than default. Set a larger estimated height to make sure autolayout sizes the view correctly
+        tableView.estimatedSectionHeaderHeight = (style == .centered) ? PhotonActionSheetUX.RowHeight : PhotonActionSheetUX.HeaderFooterHeight
+        tableView.isScrollEnabled = true
+        tableView.showsVerticalScrollIndicator = false
+        tableView.layer.cornerRadius = PhotonActionSheetUX.CornerRadius
+        tableView.separatorStyle = .none
+        tableView.cellLayoutMarginsFollowReadableWidth = false
+        tableView.accessibilityIdentifier = "Context Menu"
+        let origin = CGPoint(x: 0, y: 0)
+        let size = CGSize(width: tableView.frame.width, height: PhotonActionSheetUX.Padding)
+        let footer = UIView(frame: CGRect(origin: origin, size: size))
+        tableView.tableHeaderView = footer
+        tableView.tableFooterView = footer.clone()
```

Comment:
Is this clone function really necessary? I know this code is ported over from Firefox but it just seems really inefficient to clone something. Does it not cause performance issues?

Reply:
It doesn't negatively impact performance any more than recreating the view line by line would. I think it's cleaner than recreating line by line.